File: pascal/simple.pas

library simple;

uses stend_interface, strings;

var

	src, dst : matrix;

function init(module_name : p_char; module_name_len : unsigned; param_num : p_unsigned; return_value_num : p_unsigned) : int; cdecl;
begin

	strcopy(module_name, 'simple_pascal');
	param_num^ := 1;
	return_value_num^ := 1;
	
	init := 0;

end;

function reset : int; cdecl;
begin
	
	reset := 0;

end;

function destroy : int; cdecl;
begin
	
	destroy := 0;

end;

function get_name(is_param : boolean; ind : unsigned; name : p_char; name_len : unsigned) : int; cdecl;
begin
	
	if is_param then
		
		strcopy(name, 'Source image')
	
	else
		
		strcopy(name, 'Destination image');
	
	get_name := 0;

end;

function get_type(is_param : boolean; ind : unsigned; tp : p_int) : int; cdecl;
begin
	
	tp^ := MATRIX_TYPE;
	get_type := 0;

end;

function get_value(is_param : boolean; ind : unsigned; value : p_void) : int; cdecl;
begin
	
	if is_param then
		
		p_matrix(value)^ := src
	
	else
		
		p_matrix(value)^ := dst;
	
	get_value := 0;

end;

function set_value(is_param : boolean; ind : unsigned; value : p_void) : int; cdecl;
begin
	
	if is_param then
		
		src := p_matrix(value)^
	
	else
		
		dst := p_matrix(value)^;
	
	set_value := 0;

end;

function run : int; cdecl;
var
	
	ctx : context;

begin
	
	ctx := context_get();
	map_set_state(@ ctx^.map_state);
	dst := map_get_frame();

	run := 0;

end;

	exports

		matrix_create				name	'matrix_create',
		matrix_delete				name	'matrix_delete',
		matrix_copy					name	'matrix_copy',
		matrix_load_image			name	'matrix_load_image',
		matrix_save_image			name	'matrix_save_image',
		matrix_get_value			name	'matrix_get_value',
		matrix_set_value			name	'matrix_set_value',
		matrix_height				name	'matrix_height',
		matrix_width				name	'matrix_width',
		matrix_number_of_channels	name	'matrix_number_of_channels',
		matrix_element_type			name	'matrix_element_type',
		matrix_pointer_to_data		name	'matrix_pointer_to_data',
		matrix_pointer_to_row		name	'matrix_pointer_to_row',  

		map_set_state				name	'map_set_state',
		map_get_state				name	'map_get_state',
		map_get_frame				name	'map_get_frame',

		context_get					name	'context_get', 
		context_create_variable		name	'context_create_variable',
		context_set_value			name	'context_set_value',
		context_get_value			name	'context_get_value',

		combine_quality				name	'combine_quality',
		skeleton					name	'skeleton',

		init						name	'init',
		reset						name	'reset',
		destroy						name	'destroy',
		run							name	'run',
		get_name					name	'get_name',
		get_type					name	'get_type',
		get_value					name	'get_value',
		set_value					name	'set_value';

end.

File: cpp/simple.cpp

#define SHARED_OBJECT

#include <cstdio>
#include <cstring>
#include <cstdint>
#include <cstdlib>

#include "interface.h"

using namespace std;

matrix src, src_thr, dst;
int hor_row = 100;

bool is_correct_index(const bool is_param, const unsigned ind)
{
	return ((is_param && ind < 3) || (! is_param && ind < 1));
}

// ############################################################################ 

int EXPORT_FUNCTION init(char * module_name, const unsigned module_name_len, unsigned * param_num, unsigned * return_value_num)
{
	src = NULL;
	src_thr = NULL;
	dst = NULL;

	strncpy(module_name, "simple", module_name_len);
	module_name[module_name_len - 1] = '\0';

	* param_num = 3;
	* return_value_num = 1;

	return 0;
}

int EXPORT_FUNCTION destroy()
{
	return 0;
}

int EXPORT_FUNCTION reset()
{
	return 0;
}

// ############################################################################ 

int EXPORT_FUNCTION get_name(const bool is_param, const unsigned ind, char * name, const unsigned name_len)
{
	if(! is_correct_index(is_param, ind))
		return -1;

	switch(ind)
	{
		case 0:
		{
			strncpy(name, is_param ? "Source image" : "Destination image", name_len);

			break;
		}
		case 1:
		{
			strncpy(name, "Source image - threshold", name_len);

			break;
		}
		case 2:
		{
			strncpy(name, "Rows number", name_len);

			break;
		}
	}

	return 0;
}

int EXPORT_FUNCTION get_type(const bool is_param, const unsigned ind, int * type)
{
	if(! is_correct_index(is_param, ind))
		return -1;

	switch(ind)
	{
		case 0:
		{
			* type = MATRIX_TYPE;

			break;
		}
		case 1:
		{
			* type = MATRIX_TYPE;

			break;
		}
		case 2:
		{
			* type = INT_TYPE;

			break;
		}
	}

	return 0;
}

int EXPORT_FUNCTION get_value(const bool is_param, const unsigned ind, void * value)
{
	if(! is_correct_index(is_param, ind))
		return -1;

	switch(ind)
	{
		case 0:
		{
			* (matrix *) value = is_param ? src : dst;

			break;
		}
		case 1:
		{
			* (matrix *) value = src_thr;

			break;
		}
		case 2:
		{
			* (int *) value = hor_row;

			break;
		}
	}

	return 0;
}

int EXPORT_FUNCTION set_value(const bool is_param, const unsigned ind, const void * value)
{
	if(! is_correct_index(is_param, ind))
		return -1;

	switch(ind)
	{
		case 0:
		{
			if(is_param)
				src = * (matrix *) value;
			else
				dst = * (matrix *) value;

			break;
		}
		case 1:
		{
			src_thr = * (matrix *) value;

			break;
		}
		case 2:
		{
			hor_row = * (int *) value;

			break;
		}
	}

	return 0;
}

// ############################################################################ 

int EXPORT_FUNCTION run()
{
#if 0

	uint8_t value;
	unsigned v, u, t, height, width, ch_num;
	
	matrix_height(src, & height);
	matrix_width(src, & width);
	matrix_number_of_channels(src, & ch_num);
	
	dst = matrix_copy(src, 1);

	for(v = hor_row; v < height; v++)
		for(u = 0; u < width; u++)
			for(t = 0; t < ch_num; t++)
			{
				matrix_get_value(src_thr, v, u, t, & value);
				matrix_set_value(dst, v, u, t, & value);
			}

#else

	context ctx = context_get();

	printf("x = %lf y = %lf h = %lf course = %lf roll = %lf pitch = %lf aspect_x = %lf aspect_y = %lf coord_system = %d frame_ind = %u frame_num = %u src_video_fname = [%s] dst_video_fname = [%s] map_fname = [%s] height_map_fname = [%s]\n",
			ctx->map_state.x, ctx->map_state.y, ctx->map_state.h, ctx->map_state.course, ctx->map_state.roll, ctx->map_state.pitch, ctx->map_state.aspect_x, ctx->map_state.aspect_y, ctx->map_state.coord_system,
			ctx->frame_ind, ctx->frame_num, ctx->src_video_fname, ctx->dst_video_fname, ctx->map_fname, ctx->height_map_fname);

	ctx->map_state.h = 1200;
	ctx->map_state.course = 0;
	ctx->map_state.roll = 0;
	ctx->map_state.pitch = -90;

	map_set_state(& ctx->map_state);
 	// 62, 46
 	// 69.1, 47.3

	dst = map_get_frame();

	printf("0x%X\n", dst);

#endif

	return 0;
}

